### PR TITLE
Remove restrictions for aliases from versions and plugins for Version catalog

### LINKED ISSUE #18167: In version catalog libs, alias 'jme3-plugins' is not a valid alias — Works in 7.1.1, fails in 7.2
When using a version catalog, the dependency [jme3-plugins](https://mvnrepository.com/artifact/org.jmonkeyengine/jme3-plugins) can be declared in `libs.versions.toml` like so:

```toml
[versions]
jmonkeyengine = "3.4.0-stable"

[libraries]
jme3-plugins = { module = "org.jmonkeyengine:jme3-plugins", version.ref = "jmonkeyengine" }
```

And then in `build.gradle`:

```groovy
dependencies {
    implementation libs.jme3.plugins
}
```

### Expected Behavior
The dependency should resolve successfully in Gradle 7.1.1 and 7.2 because the alias `plugins` is *not* [a reserved alias name](https://docs.gradle.org/7.2/userguide/version_catalog_problems.html#reserved_alias_name) in 7.2 (although the singular `plugin` is).

If the plural `plugins` is supposed to be a reserved alias, then the documentation should reflect that.

### Current Behavior
In Gradle 7.1.1, the dependency resolves successfully and the project can be compiled.

In Gradle 7.2, the build fails with the following error:

```
org.gradle.api.InvalidUserDataException: Invalid catalog definition:
  - Problem: In version catalog libs, alias 'jme3-plugins' is not a valid alias.
```

### Context

<!--- In the rare cases where this is infeasible, we will also accept a detailed set of instructions. -→
1. Unzip [Jme3Plugins.zip](https://github.com/gradle/gradle/files/7087122/Jme3Plugins.zip)
2. From the root directory, run `gradlew build`.

It is currently set to wrapper 7.2 and fails to build with the above error. 

If you change it to wrapper 7.1.1, it will successfully build.

### PR DESCRIPTION
**Pull request by asodja**
(Updated)
Removed restrictions for plugins and versions since they are accessed with:
`libs.plugins.<name-of-plugin>` and `libs.versions.<name of version>`.

For dependencies:
We still apply restriction but we apply them only for first part of alias for:
- bundles
- versions
- plugins

So for example:
`plugins-my,  bundles-my` are forbidden. But `my-bundles` is allowed.

Also all camelCase cases works now without any issue.

I removed VersionCatalog interface from generated classes, so there is no more `findDependency&friends` on generated accessors. This is good for lifting restrictions, but my break some people's usages. But users can still access VersionCatalog API through VersionCatalogExtension.

----
I also have to update documentation, will do this after https://github.com/gradle/gradle/pull/18207 is merged. And since I added word `dependency` and `dependencies` I will add this to potential breaking changes

Fixes #18167

### CODE REVIEW

**Comment by asodja** [comment]
@bot-gradle test QFL

**Comment by asodja** [comment]
@bot-gradle test QFL

**Comment by asodja** [comment]
Thinking about it, maybe would make sense to move all optional API methods to some "namespace", like api, e.g. `libs.api.findDependency()`. 

And then restricted words would be `api, bundles, plugins, versions` without any need to do any of this `contains('version')` magic. WDYT?

**Comment by asodja** [comment]
@bot-gradle test QFL

**Comment by asodja** [comment]
@bot-gradle test QFL

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultMinimalDependency.java [hunk 1]
```diff
@@ -72,6 +72,9 @@ private int doComputeHashCode() {
 
     @Override
     public String toString() {
-        return module + ":" + versionConstraint;
+        String versionConstraintAsString = versionConstraint.toString();
```
Comment (ljacomet) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultMinimalDependency.java:75]: Why did you do this and the similar change in `DefaultPluginDependency`?
I am a bit concerned about potential side effects.

**Review by ljacomet — COMMENTED** [review]
One question around a `toString` change:

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultMinimalDependency.java (hunk 1 above)
Comment (asodja) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dependencies/DefaultMinimalDependency.java:75]: I noticed that if you use `withoutVersion` in version catalog, then toString returns `group-id:artifact-id:` (note the `:` at the end), with this change it returns `group-id:artifact-id`. 

ToString method for this class and `DefaultPluginDependency` was added in https://github.com/gradle/gradle/pull/18218 so it should not break anything.

**Review by ljacomet — APPROVED** [review]

**Comment by asodja** [comment]
@bot-gradle test and merge

**Pull request merged**